Adding glif_psc_double_alpha model

shixnya opened the pull request:
An extension of the glif_psc. This model uses sum of two alpha functions instead of a single alpha function as the post synaptic current input.

heplesser reviewed (changes requested):
I just put in a formal block until formalities are in place.

janskaar reviewed (changes requested):
Since this is the only model with this type of postsynaptic current, should there perhaps also be included a test of its correctness?

shixnya commented:
> Since this is the only model with this type of postsynaptic current, should there perhaps also be included a test of its correctness?

Hi janskaar,

Thank you for the code review. I want to make a file for testing the synaptic connection of this model. Where do you think is the most appropriate place for it? Will it be 'testsuite/pytests/sli2py_neurons' or 'testsuite/unittests/?

janskaar commented:
> > Since this is the only model with this type of postsynaptic current, should there perhaps also be included a test of its correctness?
> 
> Hi janskaar,
> 
> Thank you for the code review. I want to make a file for testing the synaptic connection of this model. Where do you think is the most appropriate place for it? Will it be 'testsuite/pytests/sli2py_neurons' or 'testsuite/unittests/?

Hi, thanks for your reply.

The tests can be placed in testsuite/pytests/

heplesser reviewed (approved):
The formalities are in place now, so I approve concerning the formalities. The usual code review should continue.

shixnya commented on models/glif_psc_double_alpha.h:
```diff
@@ -0,0 +1,482 @@
+tau_syn                    double vector  Rise time constants of the synaptic
+                                          alpha function in ms
+tau_syn_slow               double vector  Rise time constants of the slower
+                                          synaptic alpha function in ms
```
Thanks! This is taken care in the new commit.

shixnya commented on models/glif_psc_double_alpha.h:
```diff
@@ -0,0 +1,482 @@
+    int refractory_steps_;             //!< Number of refractory steps remaining
+    std::vector< double > y1_;         //!< synapse current evolution state 1 in pA
+    std::vector< double > y2_;         //!< synapse current evolution state 2 in pA
+    std::vector< double > y1_slow_;    //!< synapse current evolution state 1 in pA
```
Taken care in the new commit.

shixnya commented on models/glif_psc_double_alpha.h:
```diff
@@ -0,0 +1,482 @@
+    std::vector< double > asc_refractory_decay_rates_; //!< after spike current decay rates during refractory
+    double phi;                                        //!< threshold voltage component coefficient
+
+    std::vector< double > P11_; //!< synaptic current evolution parameter
```
Taken care in the new commit.

shixnya commented:
I made change to my PR so that all the parameters associated with the fast synaptic components are marked with '_fast'. I also added a file for checking some properties of the double alpha synapses (based on the test for glif_psc model), and fixed two other files that required extra parameters for testing the 'glif_psc_double_alpha' model.

clinssen commented on models/glif_psc_double_alpha.h:
```diff
@@ -0,0 +1,481 @@
+(GLIF) models [1]_ with alpha-function shaped synaptic currents.
+Incoming spike events induce a postsynaptic change of current modeled
+by sum of two alpha functions [2]_. The alpha function is normalized such that an event
+of weight 1.0 results in a peak current of the fast component of the alpha function to
```
sentences got mixed up here

shixnya replied:
Hi clinssen,

Thank you for your review. Please see my attempt to fix the description (I just pushed the changes). Each receptor gets two alpha function inputs at different time constants. Because these are alpha functions, there is only one time constant for each (Rise = Decay). You can define multiple receptor types by making the parameters to be arrays, but each receptor has exactly two alpha functions.

I wasn't sure how the math notation works, so I put down an equation, but it may not render correctly. I appreciate it if you could point me to a reference for formatting specifications.

clinssen commented on models/glif_psc_double_alpha.h:
```diff
@@ -0,0 +1,481 @@
+                                          synaptic alpha function in ms
+tau_syn_slow               double vector  Rise time constants of the slower
+                                          synaptic alpha function in ms
+amp_slow                   double vector  Relative amplitude of the slower
```
Sorry, but I find it a bit hard to interpret the meaning of the parameters. Could you perhaps express the double-alpha function mathematically?

Something like, "there are as many alpha functions as ``len(amp_slow)``; then, I_syn_total = \sum_i alpha_function[tau_rise = tau_rise_slow[i], tau_decay = tau_decay_slow[i]] + alpha_function[tau_rise = tau_rise_fast[i], tau_decay = tau_decay_fast[i]]". Thanks!

shixnya replied:
Please see my comment above. I put down an equation in the Description in the file and changed the 'Rise time constants' to 'Time constants' because, for alpha functions, there is only one time constant. I made the same change to the original glif_psc.h as well.

clinssen replied:
Great, thank you!

clinssen commented on models/glif_psc_double_alpha.h:
```diff
@@ -44,16 +44,20 @@ Description
+This function is normalized such that an event of weight 1.0 results in a peak current
+of the fast component of the alpha function to be 1 pA at :math:`t = tau_syn_fast`.
+The relative peak current of the slow component is given as amp_slow, at
+:math:`t = tau_syn_slow`. Namely,
```
Somehow Github is not allowing me to make inline suggestions.

An underscore in LaTeX means subscript, so I would suggest ``:math:`t = \tau_\text{syn\_slow}` ``. Same for the alpha functions (``\text{alpha\_function}``).

shixnya replied:
Thanks! I edited newly added equations LaTeX compatible.

It looks like the documentation of the old glif_psc.h has some equations that are not properly formatted (I assume that \mathrm is not recognized somehow). Should I work on these as well in this opportunity?
https://nest-simulator.readthedocs.io/en/stable/models/glif_psc.html

shixnya replied:
Great! Thank you for letting me know!

clinssen reviewed (approved):
Apologies, the underscore should not be prefixed when it's in a \text{} block. I made some suggestions to fix the rendering.

As these are just some small changes, I am already approving this PR. Many thanks!

I would recommend to tackle the issues with glif_psc.h in a separate PR to keep each individual PR lean.

nicolossus commented on models/glif_psc_double_alpha.h:
```diff
@@ -0,0 +1,487 @@
+Short description
++++++++++++++++++
+
+Current-based generalized leaky integrate-and-fire (GLIF) models
+(from the Allen Institute)
```
```suggestion
Current-based generalized leaky integrate-and-fire (GLIF) models with double alpha-function
(from the Allen Institute)
```

The short description is used as a unique name when referencing the model in the documentation. It should therefore be different from the one in `glif_psc.h`.

nicolossus commented on models/glif_psc_double_alpha.h:
```diff
@@ -0,0 +1,487 @@
+by the sum of two alpha functions (fast and slow components) for each receptor [2]_.
+This function is normalized such that an event of weight 1.0 results in a peak current
+of the fast component of the alpha function to be 1 pA at
+:math:`t = \tau_\text{syn_fast}`.
```
```suggestion
:math:`t = \tau_\text{syn, fast}`.
```
I think a comma instead of a underscore separating the subscript text will look better here from a mathematical notation viewpoint. Same applies for the similar ones below.

nicolossus commented on models/glif_psc_double_alpha.h:
```diff
@@ -0,0 +1,487 @@
+This function is normalized such that an event of weight 1.0 results in a peak current
+of the fast component of the alpha function to be 1 pA at
+:math:`t = \tau_\text{syn_fast}`.
+The relative peak current of the slow component is given as amp_slow, at
```
```suggestion
The relative peak current of the slow component is given as ``amp_slow``, at
```
Make model parameters more distinguishable by highlighting.

nicolossus commented on models/glif_psc_double_alpha.h:
```diff
@@ -0,0 +1,487 @@
+:math:`t = \tau_\text{syn_fast}`.
+The relative peak current of the slow component is given as amp_slow, at
+:math:`t = \tau_\text{syn_slow}`. Namely,
+:math:`I_\text{syn} = \text{alpha_function}(\tau_\text{syn} = \tau_\text{syn_fast}) +
+\text{amp_slow} * \text{alpha_function}(\tau_\text{syn} = \tau_\text{syn_slow})`.
```
```suggestion

.. math::

    I_\text{syn} = \text{alpha_function} \left( \tau_\text{syn} = \tau_\text{syn, fast} \right) + \text{amp_slow} \cdot \text{alpha_function} \left( \tau_\text{syn} = \tau_\text{syn, slow} \right).

```
I think having this equation in a math block instead of inline makes the text less dense and improves the readability a bit.

nicolossus commented on models/glif_psc_double_alpha.h:
```diff
@@ -0,0 +1,487 @@
+:math:`t = \tau_\text{syn_slow}`. Namely,
+:math:`I_\text{syn} = \text{alpha_function}(\tau_\text{syn} = \tau_\text{syn_fast}) +
+\text{amp_slow} * \text{alpha_function}(\tau_\text{syn} = \tau_\text{syn_slow})`.
+Therefore if amp_slow is not 0, the peak current of the total synaptic current is larger
+than the specified weight. By default, glif_psc_double_alpha has a single synapse that
+is accessible through receptor_port 1. An arbitrary number of synapses with different
+time constants and amp_slow can be configured by setting the desired parameters of
+tau_syn_fast, tau_syn_slow, and amp_slow arrays. The resulting synapses are addressed
+through receptor_port 1, 2, 3, ....
```
```suggestion
Therefore if ``amp_slow`` is not 0, the peak current of the total synaptic current is larger
than the specified weight. By default, ``glif_psc_double_alpha`` has a single synapse that
is accessible through ``receptor_port`` 1. An arbitrary number of synapses with different
time constants and ``amp_slow`` can be configured by setting the desired parameters of
``tau_syn_fast``, ``tau_syn_slow``, and ``amp_slow`` arrays. The resulting synapses are addressed
through ``receptor_port`` 1, 2, 3, ....
```
Highlighting.

nicolossus commented on models/glif_psc_double_alpha.h:
```diff
@@ -0,0 +1,487 @@
+being rounded to appropriate digits for simplification.
+
+For models with spike dependent threshold (i.e., GLIF2, GLIF4 and GLIF5),
+parameter setting of voltage_reset_fraction and voltage_reset_add may lead to the
```
```suggestion
parameter setting of ``voltage_reset_fraction`` and ``voltage_reset_add`` may lead to the
```

nicolossus commented on models/glif_psc_double_alpha.h:
```diff
@@ -0,0 +1,487 @@
+       model for the large, fluctuating synaptic conductance state typical of
+       neocortical neurons in vivo. J.  Comput. Neurosci., 16, 159-175.
+
+See also
+++++++++
+
+gif_psc_exp_multisynapse, gif_cond_exp, gif_cond_exp_multisynapse, gif_pop_psc_exp,
+glif_psc
```
Can you also add a reference to ``glif_psc_double_alpha`` in the See also section of ``glif_psc.h``?

shixnya replied:
I added the reference to 'glif_psc.h'.

nicolossus reviewed (changes requested):
@shixnya Thanks for this PR! This looks overall good to me. I have a few nitpicks, however, see comments below. 

Could you also include a simple example that highlights the difference between this model and `glif_psc`?

shixnya commented:
Thank you all, for thorough reviews on the code and documentation. I think I addressed all the issues raised. Please let me know if there are any other things that need to be changed. Now I see "1 workflow awaiting approval". How can we proceed from here?

nicolossus commented:
@shixnya I also think the code and documentation is good to go. I still think an example that showcases the difference in dynamics of `glif_psc_double_alpha` and `glif_psc` should be added. 

Examples reside in the `pynest/examples/` directory.  You can for instance take a look at the `glif_psc` example ( [pynest/examples/glif_psc_neuron.py](https://github.com/nest/nest-simulator/blob/master/pynest/examples/glif_psc_neuron.py)) for how examples are structured and formatted. In order to include the example in the documentation, you need to add it to

https://github.com/nest/nest-simulator/blob/75718a8123a59cb524f5b14b3a13552b4be0f7e9/doc/htmldoc/examples/index.rst?plain=1#L74-L79

and also at the bottom of the same file

https://github.com/nest/nest-simulator/blob/75718a8123a59cb524f5b14b3a13552b4be0f7e9/doc/htmldoc/examples/index.rst?plain=1#L261-L262

We need either @clinssen or @heplesser to start the CI workflow.

shixnya commented:
@nicolossus Thank you for reminding me! Yes. I added an example file for the double alpha synapse. And reapplied clang-format for the .h file that was modified. I hope this addresses all the concerns!

nicolossus commented on pynest/examples/glif_psc_double_alpha_neuron.py:
```diff
@@ -0,0 +1,214 @@
+# You should have received a copy of the GNU General Public License
+# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+
+# %%
```
```suggestion
```

nicolossus commented on pynest/examples/glif_psc_double_alpha_neuron.py:
```diff
@@ -0,0 +1,214 @@
+Simple example of how to use the ``glif_psc_double_alpha`` neuron model that illustrates
+difference from the ``glif_psc`` neuron model.
+
+The behavior of the ``glif_psc_double_alpha`` neuron model the same as the ``glif_psc``
```
```suggestion
The behavior of the ``glif_psc_double_alpha`` neuron model is the same as the ``glif_psc``
```

nicolossus commented on pynest/examples/glif_psc_double_alpha_neuron.py:
```diff
@@ -0,0 +1,214 @@
+The behavior of the ``glif_psc_double_alpha`` neuron model the same as the ``glif_psc``
+neuron model, except that the synaptic currents are modeled as a double alpha function.
+Therefore, in this example, we only compare the difference in the synaptic currents
+between the two models. compared to the single alpha function, the double alpha function
```
```suggestion
between the two models. Compared to the single alpha function, the double alpha function
```

nicolossus commented on pynest/examples/glif_psc_double_alpha_neuron.py:
```diff
@@ -0,0 +1,214 @@
+between the two models. compared to the single alpha function, the double alpha function
+has much more control over the shape of the tail of the synaptic current.
+
+Simple synaptic inputs are applied to the neuron and the resulting voltage traces,
+current traces are shown for the two models.
+
+"""
```
```suggestion
Simple synaptic inputs are applied to the neuron and the resulting voltage and
current traces are shown for the two models.
"""
```

nicolossus commented on pynest/examples/glif_psc_double_alpha_neuron.py:
```diff
@@ -0,0 +1,214 @@
+# We also pre-define the synapse time constant arrays.
+# In contrast to ``glif_psc`` models, ``glif_psc_double_alpha`` models have
+# two components of synaptic currents, one for the fast component and the other
+# for the slow component. The relative amplitude also need to be set, so there
+# are three parameters to define per port. For this example, we keep the
```
```suggestion
# for the slow component. The relative amplitude also needs to be set, so there
# are three parameters to define per receptor port. For this example, we keep the
```

nicolossus commented on pynest/examples/glif_psc_double_alpha_neuron.py:
```diff
@@ -0,0 +1,214 @@
+neurons = n_glif_psc + n_glif_psc_double_alpha_timing + n_glif_psc_double_alpha_amp
+
+###############################################################################
+# For the stimulation input to the glif_psc neurons, we create three excitation
```
```suggestion
# For the stimulation input to the ``glif_psc`` neurons, we create three excitation
```

nicolossus reviewed (approved):
@shixnya Thanks! I have some minor grammatical suggestions in the example, but overall this looks good to me so I will go ahead and approve 👍

shixnya commented:
Sorry. I accidentally closed this PR while trying to rebase to the updated upstream master.
Please let me open another PR for this.

nicolossus commented:
@shixnya I think you just can undo the rebase and reopen this PR.

shixnya commented:
@nicolossus Thanks for the comment. There is a newly added test in the updated portion of the code that requires some changes to accept the glif_psc_double_alpha model, so undoing the rebase will fail the test.

Now I see an option to 'Reopen' this PR, but somehow I see 0 changes between nest-simulator/master and shixnya/master in this page, even though mine contains all the changes I made.

Without seeing your message, I've already created a new PR. Should I proceed with it, or try to stick with this one?

nicolossus commented:
I see, I am not quite sure what the preferred option is then. We could probably proceed with the new PR, but I think we need green light from a senior developer first (@heplesser).
